File: random/fibonacci.cc
#include <iostream>
#include <cstdlib>
#include <cstring>
using namespace std;
double fib_map[5000];

double fib_rec(double n);
double fib_iter(double n);
double fib_dp(int n);

int main(int argc, char **argv)
{
	memset(fib_map, 0, sizeof(fib_map));
	fib_map[1] = 1;
	fib_map[2] = 1;

	// double num;
	// cin >> num;
	cout << "The " << argv[1] << "th fib_reconacci number is " << fib_iter(atoi(argv[1])) << endl;
}

//45th - 35 seconds
double fib_rec(double n)
{
	if (n <= 0) return -1;
	else if (n == 1) return 1;
	else if (n == 2) return 1; 
	return fib_rec(n-1) + fib_rec(n-2);
}

double fib_dp(int n)
{
	if (fib_map[n] != 0)
	{
		return fib_map[n];
	}
	
	if (fib_map[n-2] == 0)
	{
		fib_dp(n-2);
	}

	if (fib_map[n-1] == 0)
	{
		fib_dp(n-1);
	}

	fib_map[n] = fib_map[n-1] + fib_map[n-2];

	return fib_map[n];
}

//45th - 0 seconds
double fib_iter(double n)
{
	double index = 0, cur = 0, prev = 1;
	while (index < n)
	{
		double tmp = cur;
		cur += prev;
		prev = tmp;
		index++;
	}
	return cur;
}

File: random/fibonacci.py
fib_arr = [0] * 1001

def fib_dp(n):
	if n <= 0 or n > 1000:
		return "Outside allowed range"
	elif n == 1:
		return fib_arr[1]
	elif n== 2:
		return fib_arr[2]

	if fib_arr[n-1] == 0:
		fib_arr[n-1] = fib_dp(n-1)
	if fib_arr[n-2] == 0:
		fib_arr[n-2] = fib_dp(n-2)

	fib_arr[n] = fib_arr[n-1] + fib_arr[n-2]
	return fib_arr[n]
	
def fib_rec(n):
	if n <= 0 or n > 1000:
		return "Outside allowed range"
	elif n == 1:
		return 1
	elif n == 2:
		return 2
	
	return fib_rec(n-1) + fib_rec(n-2)

def fib_iter(n):
	if n <= 0 or n > 1000000:
		return "Outside allowed range"
	elif n == 1:
		return 1
	elif n == 2:
		return 1

	i = 0
	a = 0
	b = 1
	while i != n-1:
		tmp = b
		b = a + b				
		a = tmp
		i+=1
	return b

#Main
fib_arr[1] = 1
fib_arr[2] = 1

fib_n = input()
print (fib_iter(int(fib_n)))
